File: python/trabalho4.py

def jogadas_aux(
      list_jogadas:list,
      index:int, 
      list_bonus:list,
      arremeso_ante_rodada:int,
      rodada_atual:int,
      score:int 
   )->tuple[list, int]:

   if index >= len(list_jogadas):
      return score

   pinos_derrubados: int = list_jogadas[index]
   if list_bonus:
      novo_score = score + ((list_bonus[0] + 1) * pinos_derrubados )   # vai adicionar a quantidade correta de pontos 
   else:
      novo_score = score +  pinos_derrubados
   
   novo_arremeso_ante:int

   if arremeso_ante_rodada == -1: #primeira jogada
      if pinos_derrubados == 10 and rodada_atual < 10: #strike sem ser no 10
         nova_list_bonus = calcula_bonus(list_bonus[1:], 2)   #strike normal
         nova_rodada = rodada_atual + 1
         novo_arremeso_ante = -1
      else:
         nova_list_bonus = calcula_bonus(list_bonus[1:], 0)  
         nova_rodada = rodada_atual
         novo_arremeso_ante = pinos_derrubados

 
   else: #segunda jogada
      nova_rodada = rodada_atual + 1
      novo_arremeso_ante = -1

      if arremeso_ante_rodada + pinos_derrubados == 10 and rodada_atual < 10 : #spare
         nova_list_bonus = calcula_bonus(list_bonus[1:], 1)
      else:
         nova_list_bonus = calcula_bonus(list_bonus[1:], 0)  

   # não teve bônus
   return jogadas_aux(list_jogadas, index + 1, nova_list_bonus, novo_arremeso_ante,nova_rodada, novo_score)

#calculo de qual sera os bonus seguintes              
def calcula_bonus(list_bonus:list[int], opcao:int)->list[int]: #0 -> norma, 1 -> spare, 2-> strike
     if not list_bonus:
         lista_intermed = [0]
     else:
         lista_intermed = list_bonus

     if opcao == 0:
          return list_bonus
     
     elif opcao == 1: #spare 
         nova_lista = [lista_intermed[0] + 1]
         return nova_lista
     else: #strike 
          nova_lista = [lista_intermed[0] + 1, 1]
          return nova_lista
          
def calcula_pontos(lista_jogadas:list[int])->int:
    return jogadas_aux(lista_jogadas,0,[0],-1,1,0)  

def string_jogada_aux(
      list_jogadas:list[int],
      index:int, 
      arremeso_ante_rodada:int, 
      string_retorno:str,
      rodada_atual:int
   )->str:
   if index >= len(list_jogadas):
      return string_retorno

   pinos_derrubados: int = list_jogadas[index] 

   if rodada_atual < 10 :
      if arremeso_ante_rodada == -1: #primeira jogada
         if pinos_derrubados == 10:
            nova_string = string_retorno + "X _ | "
            novo_arremeso_ante = -1 #vamos para uma nova rodada depois do strike
            prox_rodada = rodada_atual + 1
         else:
            novo_arremeso_ante = pinos_derrubados
            nova_string = string_retorno + f"{pinos_derrubados} "
            prox_rodada = rodada_atual
      else: #segunda jogada
         novo_arremeso_ante = -1
         if arremeso_ante_rodada + pinos_derrubados == 10: #spare
            nova_string = string_retorno + "/ | "
         else:
            nova_string = string_retorno + f"{pinos_derrubados} | "
         prox_rodada = rodada_atual + 1
   else:
      
      if pinos_derrubados == 10:
         nova_string = string_retorno + "X "
      else:
         nova_string = string_retorno + f"{pinos_derrubados}"
      
      novo_arremeso_ante = pinos_derrubados
      prox_rodada = rodada_atual

   return string_jogada_aux(list_jogadas,index+1,novo_arremeso_ante,nova_string,prox_rodada)

def string_jogada(list_jogada:list[int])->str:
   return string_jogada_aux(list_jogada,0,-1,"",1) + "|"

jogadas:list = [ 1, 4, 4, 5, 6, 4, 5, 5, 10, 0, 1, 7, 3, 6, 4, 10, 2, 8, 6 ]
jogadas2:list = [10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10]
print(string_jogada(jogadas))
   

File: Haskel/trabalho4/trabalho4.hs
-- Define the auxiliary function for calculating scores
jogadasAux :: [Int] -> [Int] -> Int -> Int -> Int -> Int
jogadasAux [] _ _ _ score = score
jogadasAux (pinosDerrubados:xs) listBonus arremesoAnteRodada rodadaAtual score =
  let novoScore = case listBonus of
                    (b:bs) -> score + ((b + 1) * pinosDerrubados)
                    []     -> score + pinosDerrubados
      (novaListBonus, novaRodada, novoArremesoAnte)
        | arremesoAnteRodada == -1 =
            if pinosDerrubados == 10 && rodadaAtual < 10
            then case listBonus of
                   (_:bs) -> (calculaBonus bs 2, rodadaAtual + 1, -1)
                   []     -> (calculaBonus [] 2, rodadaAtual + 1, -1)
            else case listBonus of
                   (_:bs) -> (calculaBonus bs 0, rodadaAtual, pinosDerrubados)
                   []     -> (calculaBonus [] 0, rodadaAtual, pinosDerrubados)
        | arremesoAnteRodada + pinosDerrubados == 10 && rodadaAtual < 10 =
            case listBonus of
              (_:bs) -> (calculaBonus bs 1, rodadaAtual + 1, -1)
              []     -> (calculaBonus [] 1, rodadaAtual + 1, -1)
        | otherwise =
            case listBonus of
              (_:bs) -> (calculaBonus bs 0, rodadaAtual + 1, -1)
              []     -> (calculaBonus [] 0, rodadaAtual + 1, -1)
  in jogadasAux xs novaListBonus novoArremesoAnte novaRodada novoScore

-- Define the function for calculating bonuses
calculaBonus :: [Int] -> Int -> [Int]
calculaBonus [] opcao
  | opcao == 0 = []
  | opcao == 1 = [1]
  | otherwise = [1, 1]
calculaBonus (x:xs) opcao
  | opcao == 0 = x:xs
  | opcao == 1 = (x + 1):xs
  | otherwise = (x + 1):1:xs

-- Define the function for calculating total points
calculaPontos :: [Int] -> Int
calculaPontos listaJogadas = jogadasAux listaJogadas [0] (-1) 1 0

stringJogadaAux :: [Int] -> Int -> String -> Int -> String
stringJogadaAux [] arremesoAnteRodada stringRetorno rodadaAtual = stringRetorno
stringJogadaAux (pinosDerrubados:xs) arremesoAnteRodada stringRetorno rodadaAtual =
  let strPrimeiraJogada = show pinosDerrubados ++ " "
      (strStrike, strSparr, strFimRodada)
        | rodadaAtual < 10 = ("X _ | ", "/ | ", show pinosDerrubados ++ " | ")
        | otherwise = ("X ", "/ ", strPrimeiraJogada)
      (novaString, novoArremesoAnte, proxRodada)
        | arremesoAnteRodada == -1 =
            if pinosDerrubados == 10
            then (stringRetorno ++ strStrike, -1, rodadaAtual + 1)
            else (stringRetorno ++ strPrimeiraJogada, pinosDerrubados, rodadaAtual)
        | arremesoAnteRodada + pinosDerrubados == 10 =
            (stringRetorno ++ strSparr, -1, rodadaAtual + 1)
        | otherwise =
            (stringRetorno ++ strFimRodada, -1, rodadaAtual + 1)
  in stringJogadaAux xs novoArremesoAnte novaString proxRodada

stringJogada :: [Int] -> String
stringJogada listJogada = stringJogadaAux listJogada (-1) "" 1 ++ "|"


leInteiros :: String -> [Int] --função para ler vários inteiros da linha de comando
leInteiros input = map read $ words input 

main :: IO ()
main = do
  
  linhaComando <- getLine --le os numeros da linha de comando
  let numeros = leInteiros linhaComando --transforma linha lida em uma lista de números 

  putStrLn $ stringJogada numeros ++ show (calculaPontos numeros) --printa o resultado na tela
